docs/qldoc-style-guide.md

```diff
@@ -0,0 +1,184 @@
+
+## QLDoc for predicates
+
+1. Refer to all predicate parameters in the predicate documentation.
+1. Reference names, such as types and parameters, using backticks `` ` ``.
+1. Give examples of code in the target language, enclosed in ```` ``` ````  or `` ` ``.
+1. Predicates that override a single predicate don't need QLDoc, as they will inherit it.
+
+### Predicates without result
+
+1. Document predicates that do not have using a third-person verb phrase of the form ``Holds if `arg` has <property>.``.
+1. Avoid:
+   - `/**Whether ...*/`
+   - `/**"Relates ...*/`
+   - Question forms:
+     - ``/**Is `x` a foo?*/``
+     - ``/**Does `x` have a bar?*/``
+
+#### Example
+
+```ql
+/**
+ * Holds if the qualifier of this call has type `qualifierType`.
+ * `isExactType` indicates whether the type is exact, that is, whether
+ * the qualifier is guaranteed not to be a subtype of `qualifierType`.
+ */
+```
+
+### Predicates with result
+
+1. Document predicates that have a result using a third-person verb phrase of the form `Gets (a|the) <thing>`.
+1. Use "if any" if the item is usually unique, but might be missing. For example
+`Gets the body of this method, if any.`.
+1. If the predicate has more complex behaviour, for example multiple arguments are conceptually "outputs", it can be described like a predicate without a result. For example
+``Holds if `result` is a child of this expression.``.
+1. Avoid:
+   - `Get a ...`
+   - `The ...`
+   - `Results in ...`
+   - Any use of `return`
+
+#### Example
+```ql
+/**
+ * Gets the expression denoting the super class of this class,
+ * or nothing if this is an interface or a class without an `extends` clause.
+  */
+```
+
+### Deprecated predicates
+
+The documentation for deprecated predicates should be updated to emphasize the deprecation and specify what predicate to use as an alternative.
+Insert a sentence of the form `DEPRECATED: Use <other predicate> instead.` at the start of the QLDoc comment. 
+
+#### Example
+
+```ql
+/** DEPRECATED: Use `getAnExpr()` instead. */
+deprecated Expr getInitializer()
+```
+
+### Internal predicates
+
+Some predicates are internal-only declarations that cannot be made private. The documentation for internal predicates should begin with `INTERNAL: Do not use.`. 
+
+#### Example
+
+```ql
+/**
+ * INTERNAL: Do not use.
+ */
+```
+
+### Special predicates
+
+Certain special predicates should be documented consistently.
+
+- Always document `toString` as 
+  
+  ```ql
+  /**Gets a textual representation of this element.*/
+  string toString() { ... } 
+  ```
+
+- Always document `hasLocationInfo` be documented like this:
+
+  ```ql
+  /**
+   * Holds if this element is at the specified location.
+   * The location spans column `startcolumn` of line `startline` to
+   * column `endcolumn` of line `endline` in file `filepath`.
+   * For more information, see
+   * [Locations](https://help.semmle.com/QL/learn-ql/ql/locations.html).
```

Comment:
```suggestion
   * [Locations](https://help.semmle.com/QL/learn-ql/locations.html).
```
Previous link didn't work.

Reply:
Works for me, but via a redirect so it's better to use the correct URL 😸